File: Part_1_Data_Exploration.R
library(sparklyr)
library(dplyr)
library(ggplot2)
library(ggthemes)
library(psych)
library(reshape2)
library(leaflet)

## Spark Config

config <- spark_config()
config$spark.hadoop.fs.s3a.aws.credentials.provider  <- "org.apache.hadoop.fs.s3a.AnonymousAWSCredentialsProvider"
config$spark.sql.catalogImplementation <- "in-memory"
sc <- spark_connect(master = "yarn-client", config=config)

s3_link_all <-
  "s3a://ml-field/demo/flight-analysis/data/airlines_csv/*"
#  "s3a://ml-field/demo/flight-analysis/data/airlines_csv/2010.csv"

cols = list(
  FL_DATE = "date",
  OP_CARRIER = "character",
  OP_CARRIER_FL_NUM = "character",
  ORIGIN = "character",
  DEST = "character",
  CRS_DEP_TIME = "character",
  DEP_TIME = "character",
  DEP_DELAY = "double",
  TAXI_OUT = "double",
  WHEELS_OFF = "character",
  WHEELS_ON = "character",
  TAXI_IN = "double",
  CRS_ARR_TIME = "character",
  ARR_TIME = "character",
  ARR_DELAY = "double",
  CANCELLED = "double",
  CANCELLATION_CODE = "character",
  DIVERTED = "double",
  CRS_ELAPSED_TIME = "double",
  ACTUAL_ELAPSED_TIME = "double",
  AIR_TIME = "double",
  DISTANCE = "double",
  CARRIER_DELAY = "double",
  WEATHER_DELAY = "double",
  NAS_DELAY = "double",
  SECURITY_DELAY = "double",
  LATE_AIRCRAFT_DELAY = "double",
  'Unnamed: 27' = "logical"
)

spark_read_csv(
  sc,
  name = "flight_data",
  path = s3_link_all,
  infer_schema = FALSE,
  columns = cols,
  header = TRUE
)

airlines <- tbl(sc, "flight_data")

airlines %>% count()

airlines %>% sample_n(10) %>% as.data.frame

## Flights Cancelled by Carrier

cancelled_flights_by_carrier <-
  airlines %>% 
  group_by(OP_CARRIER) %>% 
  filter(CANCELLED == 1) %>%
  summarise(count_delays = n()) %>%
  arrange(desc(count_delays)) 
  #collect()

flights_by_carrier <-
  airlines %>% 
  group_by(OP_CARRIER) %>% 
  summarise(count = n()) %>%
  arrange(desc(count))
  #collect()

flights_by_carrier %>% 
  left_join(cancelled_flights_by_carrier, by = "OP_CARRIER") %>% 
  mutate(delay_percent = (count_delays/count)*100) %>%
  arrange(desc(delay_percent))



# Plot number of flights per year

## TIP
## This is important, you can run spark.sql functions inside R

flight_counts_by_year <-
  airlines %>% 
  mutate(year = year(FL_DATE)) %>%
  group_by(year) %>% 
  summarise(count = n())

cancel_counts_by_year <-
  airlines %>% 
  filter(CANCELLED == 1) %>%
  mutate(year = year(FL_DATE)) %>%
  group_by(year) %>% 
  summarise(count_delays = n())

flight_cancel_percent_year <-
  flight_counts_by_year %>% 
  left_join(cancel_counts_by_year, by = "year") %>% 
  mutate(delay_percent = (count_delays/count)*100) %>%
  arrange(desc(delay_percent))

g <- 
  ggplot(flight_cancel_percent_year, aes(year, delay_percent)) + 
  theme_tufte(base_size=14, ticks=F) + 
  geom_col(width=0.75, fill = "grey") +
  theme(axis.title=element_blank()) +
  scale_x_continuous(breaks=seq(2008,2018,1)) +
  ylab("%") +
  scale_y_continuous() + 
  ggtitle("Percentage Cancelled Flights by Year") + 
  geom_hline(yintercept=seq(1, 2, 1), col="white", lwd=0.5)
plot(g)

## Flights Cancelled by Week of Year

flight_counts_by_week <-
  airlines %>% 
  mutate(week = weekofyear(FL_DATE)) %>%
  group_by(week) %>% 
  summarise(count = n())

cancel_counts_by_week <-
  airlines %>% 
  filter(CANCELLED == 1) %>%
  mutate(week = weekofyear(FL_DATE)) %>%
  group_by(week) %>% 
  summarise(count_delays = n())

flight_cancel_percent_week <- 
  flight_counts_by_week %>% 
  left_join(cancel_counts_by_week, by = "week") %>% 
  mutate(delay_percent = (count_delays/count)*100) 

g <- 
  ggplot(flight_cancel_percent_week, aes(week, delay_percent)) + 
  theme_tufte(base_size=14, ticks=F) + 
  geom_col(width=0.75, fill = "grey") +
  theme(axis.title=element_blank()) +
  scale_x_continuous(breaks=seq(1,54,2)) +
  ylab("%") +
  scale_y_continuous() + 
  ggtitle("Percentage Cancelled Flights by Week of Year") + 
  geom_hline(yintercept=seq(1, 4, 1), col="white", lwd=0.5)
plot(g)

## Flights cancelled by Route, for both directions ORIG<>DEST

all_routes <- airlines %>% 
  mutate(combo_hash= hash(ORIGIN) + hash(DEST),combo = paste(ORIGIN,DEST,sep="")) %>% 
  select(combo_hash, combo,ORIGIN, DEST) %>%
  group_by(combo_hash) %>%
  summarize(count_all = n(), first_val = first_value(combo)) %>%
  arrange(desc(count_all)) %>%
  collect

cancelled_routes_all <- airlines %>% 
  filter(CANCELLED == 1) %>%
  mutate(combo_hash= hash(ORIGIN) + hash(DEST),combo = paste(ORIGIN,DEST,sep="")) %>% 
  select(combo_hash, combo,ORIGIN, DEST) %>%
  group_by(combo_hash) %>%
  summarize(count = n(), first_val = first_value(combo)) %>%
  arrange(desc(count)) %>%
  collect
 
cancelled_routes_percentage <-
  cancelled_routes_all %>% 
  inner_join(all_routes,by="combo_hash") %>%
  mutate(
    route = paste(
          substr(first_val.x,0,3), "<>",dest = substr(first_val.x,4,6),sep = ""
        ), 
    cancelled_percent = count/count_all*100) %>% 
  select(route,count_all,count_all,cancelled_percent) %>%
  arrange(desc(cancelled_percent)) 
  
cancelled_routes_percentage %>% as.data.frame

## Cancelled Routes Plotted on a Map

spark_read_csv(
  sc,
  name = "airports",
  path = "s3a://ml-field/demo/flight-analysis/data/airports_orig.csv",
  infer_schema = TRUE,
  header = TRUE
)

airports  <- tbl(sc, "airports")

airports <- airports %>% collect


cancelled_routes_combo <- airlines %>% 
  filter(CANCELLED == 1) %>%
  mutate(combo_hash= hash(ORIGIN) + hash(DEST),combo = paste(ORIGIN,DEST,sep="")) %>% 
  select(combo_hash, combo,ORIGIN, DEST) %>%
  group_by(combo_hash) %>%
  summarize(count = n(), first_val = first_value(combo)) %>%
  arrange(desc(count)) %>%
  collect

cancelled_routes_combo <- cancelled_routes_combo %>% 
  mutate(orig = substr(first_val,0,3), dest = substr(first_val,4,6)) %>%
  select(count,orig,dest)%>%
  inner_join(airports, by=c("orig"="iata"))%>%
  mutate(orig_lat = lat, orig_long = long) %>%
  select(count,orig,dest,orig_lat,orig_long) %>% 
  inner_join(airports, by=c("dest"="iata"))%>%
  mutate(dest_lat = lat, dest_long = long) %>%
  select(count,orig,dest,orig_lat,orig_long,dest_lat,dest_long) %>% 
  filter(between(orig_lat, 20, 50),count > 500)


map3 = leaflet(cancelled_routes_combo) %>% 
  addProviderTiles(providers$CartoDB.Positron)

for(i in 1:nrow(cancelled_routes_combo)){
    map3 <- 
      addPolylines(
        map3, 
        lat = as.numeric(cancelled_routes_combo[i,c(4,6)]), 
        lng = as.numeric(cancelled_routes_combo[i,c(5,7)]),
        weight = 10*(as.numeric(cancelled_routes_combo[i,1]/9881)+0.1), 
        opacity = 0.8*(as.numeric(cancelled_routes_combo[i,1]/9881)+0.05),
        color = "#888"
      )
}
map3

## Which columns are useful?

## TIP

unused_columns <- airlines %>%
  filter(CANCELLED == 1) %>%
  summarise_all(~sum(as.integer(is.na(.)))) %>%
  select_if(~sum(.) > 0) 

unused_columns %>% as.data.frame

unused_columns %>% colnames

#  filter_all(any_vars(. > 1)) %>% as.data.frame

File: Part_1_Data_Exploration.py
from pyspark.sql import SparkSession
import os
from pyspark.sql.types import *
from pyspark.sql.functions import *
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



spark = SparkSession\
    .builder\
    .appName("Airline")\
    .config("spark.hadoop.fs.s3a.aws.credentials.provider","org.apache.hadoop.fs.s3a.AnonymousAWSCredentialsProvider")\
    .getOrCreate()

#    .config("spark.executor.memory","8g")\
#    .config("spark.executor.cores","4")\
#    .config("spark.driver.memory","6g")\
#    .config("spark.executor.instances","5")\
#    .config("spark.dynamicAllocation.enabled","false")\

# TIP
from IPython.core.display import HTML
HTML('<a href="http://spark-{}.{}">Spark UI</a>'.format(os.getenv("CDSW_ENGINE_ID"),os.getenv("CDSW_DOMAIN")))

schema = StructType([StructField("FL_DATE", TimestampType(), True),
StructField("OP_CARRIER", StringType(), True),
StructField("OP_CARRIER_FL_NUM", StringType(), True),
StructField("ORIGIN", StringType(), True),
StructField("DEST", StringType(), True),
StructField("CRS_DEP_TIME", StringType(), True),
StructField("DEP_TIME", StringType(), True),
StructField("DEP_DELAY", DoubleType(), True),
StructField("TAXI_OUT", DoubleType(), True),
StructField("WHEELS_OFF", StringType(), True),
StructField("WHEELS_ON", StringType(), True),
StructField("TAXI_IN", DoubleType(), True),
StructField("CRS_ARR_TIME", StringType(), True),
StructField("ARR_TIME", StringType(), True),
StructField("ARR_DELAY", DoubleType(), True),
StructField("CANCELLED", DoubleType(), True),
StructField("CANCELLATION_CODE", StringType(), True),
StructField("DIVERTED", DoubleType(), True),
StructField("CRS_ELAPSED_TIME", DoubleType(), True),
StructField("ACTUAL_ELAPSED_TIME", DoubleType(), True),
StructField("AIR_TIME", DoubleType(), True),
StructField("DISTANCE", DoubleType(), True),
StructField("CARRIER_DELAY", DoubleType(), True),
StructField("WEATHER_DELAY", DoubleType(), True),
StructField("NAS_DELAY", DoubleType(), True),
StructField("SECURITY_DELAY", DoubleType(), True),
StructField("LATE_AIRCRAFT_DELAY", DoubleType(), True)])


flight_df = spark.read.csv(
  path="s3a://ml-field/demo/flight-analysis/data/airlines_csv/*",header=True,
  schema=schema)

#  
  
### Soo much faster
flight_df.persist()

## Flights Cancelled by Carrier

## Tip put your filter first

cancel_by_carrier = flight_df\
  .filter("CANCELLED == 1")\
  .groupby("OP_CARRIER")\
  .count()\
  .sort("count",ascending=False)\
  .withColumnRenamed('count', 'count_delays')

  
flight_by_carrier = flight_df\
  .groupby("OP_CARRIER")\
  .count()\
  .sort("count",ascending=False)\
  .withColumnRenamed('count', 'count_total')

  
cancel_by_carrier_percent = flight_by_carrier\
  .join(
    cancel_by_carrier, 
    flight_by_carrier.OP_CARRIER == cancel_by_carrier.OP_CARRIER
  )
  
cancel_by_carrier_percent\
  .withColumn(
    "delay_percent",(
      cancel_by_carrier_percent.count_delays/cancel_by_carrier_percent.count_total
    )*100
  )\
  .sort("delay_percent",ascending=False)\
  .toPandas()


# Plot number of flights per year

## TIP
## This is important, you can run spark.sql functions inside R  

flight_by_year = flight_df\
  .withColumn("year",year("FL_DATE"))\
  .groupby("year")\
  .count()\
  .sort("count",ascending=False)\
  .withColumnRenamed('count', 'count_total')
  
cancel_by_year = flight_df\
  .filter("CANCELLED == 1")\
  .withColumn("year_cancel",year("FL_DATE"))\
  .groupby("year_cancel")\
  .count()\
  .sort("count",ascending=False)\
  .withColumnRenamed('count', 'cancel_total')

cancel_by_year_percent = flight_by_year\
  .join(
    cancel_by_year, 
    flight_by_year.year == cancel_by_year.year_cancel
  )  
  
cancel_by_year_percent = cancel_by_year_percent\
  .withColumn(
    "delay_percent",(
      cancel_by_year_percent.cancel_total/cancel_by_year_percent.count_total
    )*100
  )\
  .sort("year",ascending=False)
  
cancel_by_year_percent_pd = cancel_by_year_percent.toPandas()


def plotter():
  sns.set_style("white",{'axes.axisbelow': False})
  plt.bar( 
    cancel_by_year_percent_pd.year, 
    cancel_by_year_percent_pd.delay_percent,
    align='center', 
    alpha=0.5,
    color='#888888',
  )
  plt.grid(color='#FFFFFF', linestyle='-', linewidth=0.5, axis='y')
  plt.title(
    'Percentage Cancelled Flights by Year',
    color='grey'
  )
  plt.xticks(
    cancel_by_year_percent_pd.year,
    color='grey'
  )
  plt.yticks(color='grey')
  sns.despine(left=True,bottom=True)
plotter()



# Plot number of flights per week  

flight_by_week = flight_df\
  .withColumn("week",weekofyear("FL_DATE"))\
  .groupby("week")\
  .count()\
  .sort("count",ascending=False)\
  .withColumnRenamed('count', 'count_total')
  
cancel_by_week = flight_df\
  .filter("CANCELLED == 1")\
  .withColumn("week_cancel",weekofyear("FL_DATE"))\
  .groupby("week_cancel")\
  .count()\
  .sort("count",ascending=False)\
  .withColumnRenamed('count', 'cancel_total')

cancel_by_week_percent = flight_by_week\
  .join(
    cancel_by_week, 
    flight_by_week.week == cancel_by_week.week_cancel
  )  
  
cancel_by_week_percent = cancel_by_week_percent\
  .withColumn(
    "delay_percent",(
      cancel_by_week_percent.cancel_total/cancel_by_week_percent.count_total
    )*100
  )\
  .sort("week",ascending=False)
  
cancel_by_week_percent_pd = cancel_by_week_percent.toPandas()


def plotter():
  sns.set_style("white",{'axes.axisbelow': False})
  plt.bar( 
    cancel_by_week_percent_pd.week, 
    cancel_by_week_percent_pd.delay_percent,
    align='center', 
    alpha=0.5,
    color='#888888',
  )
  plt.grid(color='#FFFFFF', linestyle='-', linewidth=0.5, axis='y')
  plt.title(
    'Percentage Cancelled Flights by Week',
    color='grey'
  )
  plt.xticks(
    color='grey'
  )
  plt.yticks(color='grey')
  sns.despine(left=True,bottom=True)
plotter()

## TIP Aggregation in the select if no groupby

all_routes = flight_df\
  .withColumn("combo_hash", hash("ORIGIN")+hash("DEST"))\
  .withColumn("combo", concat(col("ORIGIN"),col("DEST")))\
  .groupby("combo_hash")\
  .agg(count("combo_hash").alias("count_all"),first("combo").alias("route_alias"))\
  .sort("count_all",ascending=False)

cancelled_routes_all = flight_df\
  .filter("CANCELLED == 1")\
  .withColumn("combo_hash", hash("ORIGIN")+hash("DEST"))\
  .withColumn("combo", concat(col("ORIGIN"),col("DEST")))\
  .groupby("combo_hash")\
  .agg(count("combo_hash").alias("count"),first("combo").alias("route_alias"))\
  .sort("count",ascending=False)  

cancelled_routes_percentage = cancelled_routes_all\
  .join(
    all_routes,
    cancelled_routes_all.combo_hash == all_routes.combo_hash
  )\
  .withColumn(
    "route_alias", 
    concat(
                subrcol("route_alias"),col("DEST")))

cancelled_routes_percentage <-
  cancelled_routes_all %>% 
  inner_join(all_routes,by="combo_hash") %>%
  mutate(
    route = paste(
          substr(first_val.x,0,3), "<>",dest = substr(first_val.x,4,6),sep = ""
        ), 
    cancelled_percent = count/count_all*100) %>% 
  select(route,count_all,count_all,cancelled_percent) %>%
  arrange(desc(cancelled_percent)) 
  
cancelled_routes_percentage %>% as.data.frame
